src/kixi/datastore/metadatastore/elasticsearch.clj

```diff
@@ -0,0 +1,239 @@
+
+(defn ensure-index
+  [conn]
+  (when-not (esi/exists? conn index-name)
+    (build-index conn)))
+
+(defn- get-document
+  [conn id]
+  (esd/get conn
+           index-name 
+           doc-type 
+           id
+           {:preference "_primary"}))
+
+(s/fdef update-metadata-processor
+        :args (s/cat :conn #(instance? clojurewerkz.elastisch.rest.Connection %)
+                     :update-req ::cs/file-metadata-updated))
+
+(defmulti update-metadata-processor
+  (fn [conn update-event]
+    (::cs/file-metadata-update-type update-event)))
+
+(def apply-attempts 10)
+
+(defn version-conflict
+  [resp]
+  (some
+   #(= "version_conflict_engine_exception"
+       (:type %))
+   ((comp :root_cause :error) resp)))
+
+(defn error?
+  [resp]
+  (:error resp))
+
+(defn- apply-func
+  ([conn id f]
+   (apply-func conn id f apply-attempts))
+  ([conn id f tries]
+   (let [curr (get-document conn id)]
+     (let [resp (esd/put conn
+                         index-name
+                         doc-type
+                         id
+                         (f curr) 
+                         (merge put-opts
+                                (when (:_version curr)
+                                  {:version (:_version curr)})))]
+       (if (and (version-conflict resp)
+                (pos? tries))
+         (recur conn id f (dec tries))
+         resp)))))
+
+(defn- merge-metadata
+  [conn id update]
+  (let [r (apply-func
+           conn
+           id   
+           (fn [curr]
+             (merge-with merge
+                         (:_source curr)
+                         (all-keys->es-format update))))]
+    (if (error? r)
+      (error "Unable to merge File Metadata for id: " id ". Trying to merge: " update ". Response: " r)
+      r)))
+
+(defn- cons-metadata
+  [conn id k element]
+  (let [r (apply-func
+           conn
+           id   
+           (fn [curr]
+             (update (:_source curr) (kw->es-format k) 
+                     #(cons (all-keys->es-format element) %))))]
+    (if (error? r)
+      (error "Unable to cons File Metadata for id: " id ". Trying to update: " k ". Response: " r)
+      r)))
+
+(defmethod update-metadata-processor ::cs/file-metadata-created
+  [conn update-event]
+  (let [metadata (::ms/file-metadata update-event)]
+    (info "Update: " metadata)
+    (merge-metadata
+     conn
+     (::ms/id metadata)
+     metadata)))
+
+(defmethod update-metadata-processor ::cs/file-metadata-structural-validation-checked
+  [conn update-event]
+  (info "Update: " update-event)
+  (merge-metadata
+   conn
+   (::ms/id update-event)
+   (select-keys update-event                
+                [::ms/structural-validation])))
+
+(defmethod update-metadata-processor ::cs/file-metadata-segmentation-add
+  [conn update-event]
+  (info "Update: " update-event)
+  (cons-metadata
+   conn
+   (::ms/id update-event)
+   ::ms/segmentations 
+   (::ms/segmentation update-event)))
+
+(defn response-event
+  [r]
+  nil)
```

Comment:
It's used in the handler comp. It's a dirty, but havn't had a better idea of how to represent this yet.